internal/migrations/changesets.go

```diff
@@ -132,49 +134,54 @@ func applyChangesets(ctx context.Context, app *common.App, blockCtx *common.Bloc
 		return nil // migration completed
 	}
 
-	currentHeight = lastChangeset + 1
-	if currentHeight == 0 {
-		currentHeight = startHeight
+	if lastChangeset == -1 {
+		lastChangeset = 0 // set the last changeset to 0, because migration declaration doesn't accept int64 (-1)
 	}
 
-	// Apply the changesets in the order of block heights
 	for {
-		// If the current height is greater than the migration end height, break
-		if currentHeight >= endHeight {
-			blockCtx.ChainContext.MigrationParams = nil
-			app.Service.Logger.Info(migrationCompleted, log.Int("height", currentHeight))
+		// get the metadata for the earliest height that has not been applied
+		height, prevHeight, totalChunks, chunksReceived, err := getEarliestChangesetMetadata(ctx, tx)
+		if err != nil {
+			return err
+		}
+
+		if height == -1 { // no changesets to apply
 			break
 		}
 
-		// Check if all chunks have been received for the current height, if not, break
-		totalChunks, chunksReceived, err := getChangesetMetadata(ctx, tx, currentHeight)
-		if err != nil {
-			return err
+		if height != startHeight && prevHeight != lastChangeset {
+			// If the previous changeset is not applied, break
+			break
```

Comment:
Is this normal to happen?  if the previous changeset was not applied, when would it be applied?  After it is retrieved and in some later block's end block hook?

Reply:
Yeah, it will be applied once we receive all the chunks for that block.